### PR TITLE
✨ Handle IP blocks based on VDC location

### PR DESCRIPTION
**Pull request by jriedel-ionos**
**What is the purpose of this pull request/Why do we need it?**
If you specify a different VDC ID in the IonosCloudMachine spec, IP blocks are now correctly booked in this location.

**Issue #, if available:**
#154

**Description of changes:**
* IP blocks are booked in the location specified in the IonosCloudMachine spec
* Location is written into the Machine status

**Special notes for your reviewer:**

**Checklist:**
- [ ] Documentation updated
- [x] Unit Tests added
- [ ] E2E Tests added
- [x] Includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)

### CODE REVIEW

File: internal/ionoscloud/client.go [hunk 1]
```diff
@@ -66,4 +66,6 @@ type Client interface {
 	GetRequests(ctx context.Context, method, path string) ([]sdk.Request, error)
 	// PatchNIC updates the NIC identified by nicID with the provided properties, returning the request location.
 	PatchNIC(ctx context.Context, datacenterID, serverID, nicID string, properties sdk.NicProperties) (string, error)
+	// GetDatacenterLocationByID returns the location of the data center identified by datacenterID.
+	GetDatacenterLocationByID(datacenterID string) (string, error)
```
Comment (lubedacht) [inline internal/ionoscloud/client.go:70]: Why not passing a context here?

File: internal/ionoscloud/client/client.go [hunk 2]
```diff
@@ -463,3 +463,17 @@ func validateNICParameters(datacenterID, serverID, nicID string) (err error) {
...
 	return nil
 }
+
+// GetDatacenterLocationByID returns the location of the data center identified by datacenterID.
+func (c *IonosCloudClient) GetDatacenterLocationByID(datacenterID string) (string, error) {
+	if datacenterID == "" {
+		return "", errDatacenterIDIsEmpty
+	}
+
+	datacenter, _, err := c.API.DataCentersApi.DatacentersFindById(context.Background(), datacenterID).Execute()
```
Comment (lubedacht) [inline internal/ionoscloud/client/client.go:473]: context.Background doesn't sound like a good idea

**Review by lubedacht — CHANGES_REQUESTED** [review]

**Review by lubedacht — DISMISSED** [review]

**Review by piepmatz — COMMENTED** [review]
What about using the location from the status if set instead of querying the Cloud API again?

File: internal/service/cloud/ipblock_test.go [hunk 3]
```diff
@@ -189,6 +191,12 @@ func (s *ipBlockTestSuite) validateRequest(method, state, path string, req *infr
 	s.Equal(path, req.RequestPath)
 }
 
+func (s *ipBlockTestSuite) validateSuccessfulReservation(location string, err error) {
+	s.T().Helper()
+	s.NoError(err)
+	s.Equal(exampleLocation, location)
+}
```
Comment (piepmatz) [inline internal/service/cloud/ipblock_test.go:198]: Right now this is mostly overhead: You use it only once and there the error was even checked before calling this method.
So rather just move the `s.Equal(exampleLocation, location)` check where this function is called.

**Review by piepmatz — APPROVED** [review]

**Pull request merged**